SinglePlaneDebug: output non-garbage values

astiob opened the pull request:
The local array declared inside `debug...()` ends its lifetime upon return, so by the time `set_frame_props` reads the memory, it can contain anything. In particular, in my tests, I consistently get all-zero `BoreAdjustment`.

To ensure the correct values are actually output, put them in a memory location that survives the function return. It seems most natural to have the caller provide this location.

Test binaries: https://github.com/astiob/bore/actions/runs/13777127050

Asd-g commented:
Yes, classic race condition.
With one thread and one call the stack memory is not overwritten because it's accessed right after the scope end.

Btw you can make the change minimal like this - https://pastebin.com/raw/weGKJki0

astiob commented:
FWIW I doubt this has anything to do with race conditions. The memory isn’t even shared; it’s part of the stack, of which each thread has its own. The array legally ceases existing as an object when the function returns, so the compiler is free to reuse the memory for anything else such as spillover register values (or indeed to remove the read entirely if it were to notice that it is invalid, because an invalid read is just a slower version of an absent read).

That said, I haven’t attempted a practical test with restricted threading—although my original tests (where I kept getting all zeros) only accessed a single frame anyway, so they probably were in fact single-threaded.

> Btw you can make the change minimal like this - https://pastebin.com/raw/weGKJki0

True. For some reason I felt that preserving the type information was useful, but perhaps it isn’t.

astiob committed:
SinglePlaneDebug: output non-garbage values

The local array declared inside debug...() ends its lifetime upon return,
so by the time set_frame_props reads the memory, it can contain anything
(or more accurately, the pointer dereference is invalid and the behavior
is undefined). For example, in my tests, I get all-zero BoreAdjustment.

To ensure the correct values are actually output,
put them in a memory location that survives the function return.
It seems most natural to have the caller provide this location.

astiob commented:
Force-pushed to simplify the type from pointer-to-array to a simple pointer to make the diff smaller.

quietvoid commented:
Looks fine to me, thanks.
